[1402] Extend csv export to include all fields

mooktakim opened the pull request:
### Context

https://trello.com/c/kB2QSJ2I/1402-m-extend-csv-export-to-include-all-fields

### Changes proposed in this pull request

* updated export columns to include all from google sheet
* some columns highlighted red on google sheet has been left blank as they are not available for now, can be added in future as they are added.

### Guidance to review

* From trainee list page click "Export these records"

mooktakim commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[M] Extend csv export to include all fields](https://trello.com/c/kB2QSJ2I/1402-m-extend-csv-export-to-include-all-fields)

archferns commented:
FYI, we might need to extend https://github.com/DFE-Digital/register-trainee-teachers/pull/1186 to cover the new fields.

mooktakim commented:
> FYI, we might need to extend #1186 to cover the new fields.

Fixed merge conflicts.

jhenderson commented on app/services/exports/trainee_search_data.rb:
```diff
@@ -4,8 +4,8 @@ module Exports
   class TraineeSearchData
     VULNERABLE_CHARACTERS = ["=", "+", "-", "@"].freeze
```
```suggestion
    VULNERABLE_CHARACTERS = %w[= + - @].freeze
```

jhenderson reviewed (approved):
LGTM

edwardhorsford reviewed (changes requested):
A couple issues with this:

1. Trainee url seems to be to localhost, not the environment url
2. Status - I'd expect `QTS recommended` or `EYTS recommended` or `QTS awarded` or `EYTS awarded` rather than `Qualification recommended` and `Qualification awarded`. If we don't store this it's not essential.
3. please can we remove `highest_qualification`
4. can we export values for `course_qualification_type` - eg postgrad or undergrad? I think we send this to dttp so have a way of deriving it from the route.
5. Can we export the course allocation subject?
6. I missed course name in the spec - could we add it after course code? eg the Publish string name for the course.
7. withdraw date format should probably match defer date format
8. Some data seems to be missing for course details where it's an apply draft
9. international address seems to be inserting too many returns
![Screenshot 2021-08-03 at 16 21 13](https://user-images.githubusercontent.com/2204224/128041742-0e097adb-0228-46b8-8689-7585e8e5161f.png)
10. Nationalities should not be lowercase I think?
11. Some fields we collect could contain commas - I think we need to quote some of our strings - eg degree grade other.
12. I don't think bursary value should output if trainee isn't explicitly applying for bursary
13. Early years bursary tier doesn't output helpfully - I selected Tier 1, but the output is `0`. Could we output something like `Tier 1`? 

Unsure:
* Provider id is showing as a 3 letter string - is this correct?

mooktakim commented:
@edwardhorsford 

> A couple issues with this:
> 
> 1. Trainee url seems to be to localhost, not the environment URL

This is set in the environment file, it should have the correct URL for production.

> 2. Status - I'd expect `QTS recommended` or `EYTS recommended` or `QTS awarded` or `EYTS awarded` rather than `Qualification recommended` and `Qualification awarded`. If we don't store this it's not essential.

These are the states:
- draft: Draft
- apply_draft: Apply Draft
- submitted_for_trn: Pending TRN
- trn_received: TRN received
- recommended_for_award: Qualification recommended
- withdrawn: Withdrawn
- deferred: Deferred
- awarded: Qualification awarded
- qts_recommended: QTS recommended
- qts_awarded: QTS awarded
- eyts_recommended: EYTS recommended
- eyts_awarded: EYTS awarded

> 3. please can we remove `highest_qualification`

Removed

> 4. can we export values for `course_qualification_type` - eg postgrad or undergrad? I think we send this to dttp so have a way of deriving it from the route.

Carded separately

> 5. Can we export the course allocation subject?

Added

> 6. I missed course name in the spec - could we add it after course code? eg the Publish string name for the course.

added

> 7. withdraw date format should probably match defer date format

Converted to date format

> 9. international address seems to be inserting too many returns

I have added some processing to remove the new lines, address will be on one line comma separated

> 10. Nationalities should not be lowercase I think?

Fixed

> 11. Some fields we collect could contain commas - I think we need to quote some of our strings - eg degree grade other.

The CSV generator will quote the values if there is comma in the value.

> 12. I don't think bursary value should output if trainee isn't explicitly applying for bursary

Fixed

> 13. Early years bursary tier doesn't output helpfully - I selected Tier 1, but the output is `0`. Could we output something like `Tier 1`?

Fixed

> 
> Unsure:
> 
> * Provider id is showing as a 3 letter string - is this correct?

This is what the factory generates in dev/test.

edwardhorsford commented:
> These are the states:
> 
> * draft: Draft
> * apply_draft: Apply Draft
> * submitted_for_trn: Pending TRN
> * trn_received: TRN received
> * recommended_for_award: Qualification recommended
> * withdrawn: Withdrawn
> * deferred: Deferred
> * awarded: Qualification awarded
> * qts_recommended: QTS recommended
> * qts_awarded: QTS awarded
> * eyts_recommended: EYTS recommended
> * eyts_awarded: EYTS awarded

I'm not seeing `qts_recommended` or `eyts_recommended` or `qts_awarded` or `eyts_awarded`. Only the more generic `qualification_recommended` or `qualification_awarded` - ideally what's exported would  match what we see when we view the record. The two types aren't mutually exclusive - but on the basis of what I see from the export, I don't know how we'd ever get the more specific ones.

As an example, the record with id `15/16-344` is in the state `QTS recommended` on the site, but when exported says `qualification_recommended`.

Other than that, all looking good 👍

mooktakim commented:
@edwardhorsford 
Status should now be same as the views.
